lib/engine-native/src/serialize.rs

```diff
@@ -33,58 +36,84 @@ pub struct ModuleMetadata {
-    pub prefix: &'a String,
+#[derive(MemoryUsage)]
+pub struct ModuleMetadataSymbolRegistry {
+    pub prefix: String,
 }
 
-impl ModuleMetadata {
-    pub fn split<'a>(
-        &'a mut self,
-    ) -> (&'a mut CompileModuleInfo, ModuleMetadataSymbolRegistry<'a>) {
-        let compile_info = &mut self.compile_info;
-        let symbol_registry = ModuleMetadataSymbolRegistry {
-            prefix: &self.prefix,
-        };
-        (compile_info, symbol_registry)
+cfg_if::cfg_if! {
+    if #[cfg(target_endian = "big")] {
+        const HOST_ENDIAN: u8 = b'b';
+    } else if #[cfg(target_endian = "little")] {
+        const HOST_ENDIAN: u8 = b'l';
     }
+    else {
+        compile_error!("Endian not supported by the host");
+    }
+}
 
-    pub fn get_symbol_registry<'a>(&'a self) -> ModuleMetadataSymbolRegistry<'a> {
+impl ModuleMetadata {
+    pub fn get_symbol_registry(&self) -> ModuleMetadataSymbolRegistry {
         ModuleMetadataSymbolRegistry {
-            prefix: &self.prefix,
+            prefix: self.prefix.clone(),
         }
     }
 
-    pub fn serialize(&mut self) -> Result<Vec<u8>, CompileError> {
+    /// Serialize the Metadata into bytes
+    /// The bytes will have the following format:
+    /// RKYV serialization (any length) + POS (8 bytes) + Endian (1 byte)
+    pub fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
         let mut serializer = SharedSerializerAdapter::new(WriteSerializer::new(vec![]));
-        let pos = serializer.serialize_value(self).map_err(to_compile_error)? as u64;
+        let pos = serializer
+            .serialize_value(self)
+            .map_err(to_serialize_error)? as u64;
         let mut serialized_data = serializer.into_inner().into_inner();
         serialized_data.extend_from_slice(&pos.to_le_bytes());
-        if cfg!(target_endian = "big") {
-            serialized_data.extend_from_slice(&[b'b']);
-        } else if cfg!(target_endian = "little") {
-            serialized_data.extend_from_slice(&[b'l']);
-        }
+        serialized_data.extend_from_slice(&[HOST_ENDIAN]);
         Ok(serialized_data)
     }
 
+    /// Deserialize the Metadata from a slice.
+    /// The slice must have the following format:
+    /// RKYV serialization (any length) + POS (8 bytes) + Endian (1 byte)
+    ///
+    /// # Safety
+    ///
+    /// This method is unsafe since it deserializes data directly
+    /// from memory.
+    /// Right now we are not doing any extra work for validation, but
+    /// `rkyv` has an option to do bytecheck on the serialized data before
+    /// serializing (via `rkyv::check_archived_value`).
     pub unsafe fn deserialize(metadata_slice: &[u8]) -> Result<Self, DeserializeError> {
         let archived = Self::archive_from_slice(metadata_slice)?;
         Self::deserialize_from_archive(archived)
     }
 
+    /// # Safety
+    ///
+    /// This method is unsafe.
+    /// Please check `ModuleMetadata::deserialize` for more details.
```

Comment:
"The `metadata_slice` must be an archive produced by ModuleMetadata::serialize"?

Reply:
Yeah, that's better!